pkg/maps/ctmap/ctmap.go

```diff
@@ -510,6 +517,108 @@ func GC(m *Map, filter *GCFilter) int {
+//
+// PurgeOrphanNATEntries() is triggered by the datapath via the GC signaling
+// mechanism. When the datapath SNAT fails to find free mapping after
+// SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring buffer.
+// The consumer of the buffer invokes the function.
+//
+// The SNAT is being used for the following cases:
+// 1. By NodePort BPF on an intermediate node before fwd'ing request from outside
+//     to a destination node.
+// 2. A packet from local endpoint sent to outside (BPF-masq).
+// 3. A packet from a host local application (i.e. running in the host netns)
+//    This is needed to prevent SNAT from hijacking such connections.
+// 4. By DSR on a backend node to SNAT responses with service IP+port before
+//    sending to a client.
+//
+// In the case of 1-3, we always create a CT_EGRESS CT entry. This allows the
+// CT GC to remove corresponding SNAT entries. In the case of 4, will create
+// CT_INGRESS CT entry. See the unit test TestOrphanNatGC for more examples.
+//
+// The function only handles 1-3 cases, the 4. case is TODO(brb).
+func PurgeOrphanNATEntries(ctMap *Map) *NatGCStats {
+	if option.Config.NodePortMode == option.NodePortModeDSR ||
+		option.Config.NodePortMode == option.NodePortModeHybrid {
+		return nil
+	}
+
+	natMap := mapInfo[ctMap.mapType].natMap
+	if natMap == nil {
+		return nil
+	}
+	stats := newNatGCStats(natMap)
+
+	cb := func(key bpf.MapKey, value bpf.MapValue) {
+		natKey := key.(*nat.NatKey4)
+		natVal := value.(*nat.NatEntry4)
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 1 { // natKey is r(everse)tuple
+			ctKey := &tuple.TupleKey4Global{TupleKey4: tuple.TupleKey4{
+				// Workaround #5848
+				SourceAddr: natKey.SourceAddr,
+				DestPort:   natKey.SourcePort,
+				DestAddr:   natVal.Addr,
+				SourcePort: natVal.Port,
+				NextHeader: natKey.NextHeader,
+				Flags:      tuple.TUPLE_F_OUT,
+			}}
+			if _, err := ctMap.Lookup(ctKey); errors.Is(err, unix.ENOENT) {
+				// No CT entry is found, so delete SNAT for both original and
+				// reverse flows
+				oNatKey := &nat.NatKey4{TupleKey4Global: tuple.TupleKey4Global{TupleKey4: tuple.TupleKey4{
+					SourceAddr: natVal.Addr,
+					SourcePort: natVal.Port,
+					DestAddr:   natKey.SourceAddr,
+					DestPort:   natKey.SourcePort,
+					NextHeader: natKey.NextHeader,
+					Flags:      tuple.TUPLE_F_OUT,
+				}}}
+				if err := natMap.Delete(natKey); err == nil {
+					stats.IngressDeleted += 1
+				}
+				if err := natMap.Delete(oNatKey); err == nil {
+					stats.EgressDeleted += 1
+				}
+			} else {
+				stats.IngressAlive += 1
+			}
+		} else { // natKey is o(riginal)tuple
+			// Check for reverse flow NAT entry. If not found, remove natKey.
+			// Note, that in the datapath otuple is inserted into the map after
+			// rtuple. This means that if we check first for rtuple and remove
+			// otuple, then there would be a race window during which to-be
+			// inserted NAT tuples by the datapath could be destroyed by the GC.
```

Comment:
I think this is handling the case:

1. Datapath creates CT entry
1. Datapath creates SNAT reverse tuple entry
1. The CT entry is deleted for some reason (?)
1. Userspace dumps the SNAT reverse tuple entry, decides to delete the SNAT reverse tuple
1. Datapath creates the SNAT forward tuple
1. Now there's an original-dir tuple but no reverse tuple.

For a given new connection, I would expect the datapath is going through steps 1,2,5 in quick succession, so I don't understand how step 3 gets interleaved. It's a new connection, so presumably it won't be timed out. Unless you're cycling the entire CT map within the time of handling a single packet, LRU wouldn't evict the new entry.

If the above is the case, it seems like this entire alternate branch is unnecessary optimization.

Is there another case you have in mind for this branch?

Reply:
(from discussion with @jrfastab out-of-band)

Reply:
Cross-posting @jrfastab question from the backport PR to keep everything in the single place:

> @brb I'm trying to understand the if/else branches here. I understand the flow of CT and SNAT map population as follows,
>
>    Create a CT entry
>   Create SNAT reply entry (TUPLE_F_IN) ->
>    Create SNAT original dir entry (TUPLE_F_OUT)
>
> My understanding is then for most cases we should have a key in both directions. At some point the CT entry gets deleted and we have an orphaned SNAT entry that needs to be deleted.
>
> At this point we start walking our nat table. If the key has TUPLE_F_IN set AND no CT Entry we will delete both oNatKey and rNatKey. All good as far as I read it.
>
> Now if the key has TUPLE_F_OUT set we use the else branch and check if its rNatKey exists. If it does leave it alone. If not its somehow been orphaned by its pair. Couple questions here.
>
>    This case where rNatKey of a TUPLE_F_OUT does not exist should be rare correct?
>  It appears to rely on a race condition. Are there other TUPLE_F_* types where it might be more common?
>
> Now working on my assumption its rare, why bother with the else branch at all? Just check for CT in both  > TUPLE_F_IN and TUPLE_F_OUT cases and if it does not exist delete the pair of keys as done in the if branch. > In the race condition handled now in the else branch it would simply fail the delete, no harm done. And by it > being rare who cares if we rarely do an extra delete on the map.
>
> Did I miss something? Seems cleaner to just handle the single case and not have an if/else here if possible.
>
> And just to be clear, I'm not arguing it is incorrect as is, but do want to be sure I understand it correctly.

Reply:
To answer to both, I thought about handling the `TUPLE_F_OUT` for cases when `TUPLE_F_IN` entries get evicted by the LRU mechanism. However, in the end I think that it's unnecessary optimization, as the datapath should handle this:
```
static __always_inline int __snat_update(void *map, void *otuple, void *ostate,
					 void *rtuple, void *rstate)
{
	int ret = map_update_elem(map, rtuple, rstate, BPF_NOEXIST);
	if (!ret) {
		ret = map_update_elem(map, otuple, ostate, BPF_NOEXIST);
		if (ret)
			map_delete_elem(map, rtuple);
	}
	return ret;
}
```
I.e. if we end up having a `TUPLE_F_IN` entry w/o a corresponding `TUPLE_F_OUT`, next time a packet matches `TUPLE_F_IN`, the `__snat_update()` will delete it at the cost of dropping the packet. IMO this cost is lower than having more expensive GC.

Thanks for pointing this out, I'm going to remove the branch.